doc/developer/design/20210715_table_update.md

```diff
@@ -0,0 +1,52 @@
+# UPDATE and DELETE for tables
+
+## Summary
+
+Support UPDATE and DELETE for tables.
+These cannot be used in transactions and are serialized along with INSERTs.
+
+## Goals
+
+Add a framework for supporting transactional table operations that need to do writes based on the result of a read.
+This framework is designed to be serializable, and so is safe when used by multiple concurrent SQL sessions.
+
+## Non-Goals
+
+General transactions with interleaved reads and writes are not a goal.
+High performance under heavy concurrent load is not a goal.
+
+## Description
+
+UPDATE, DELETE, UPSERT, INSERT ON CONFLICT, and unique key constraints are similar operations: a read followed by a write depending on the results of the read:
+- DELETE will remove all read rows
+- UPDATE will remove all read rows and add a new row for each removed
+- unique key constraints will (during UPDATE and INSERT) do a read looking for any rows that might already exist, and do nothing (or error) if there are any result rows
+- UPSERT and INSERT ON CONFLICT will do the same as unique key constraints above, but will instead not error and either do nothing for each conflict row or remove it and add a new one
+
+We can add basic support for UPDATE and DELETE with a new plan node called ReadThenWrite that can be configured to do a read and then arbitrary behavior on its results.
+We can add support for the others at a later date.
+
+ReadThenWrite is a node that performs a Peek at some time `t1`.
+Once the Peek has results ready, they are read and a set of diffs is produced which are sent at some time `t2` where `t1 > t2`.
+
+In order to support our goal of strict serializability, we need to ensure that no writes to tables affected by ReadThenWrite happen between `t1` and `t2`.
```

Comment:
Should the previous line be t1 < t2, or should this also state that "we need to ensure that no reads occur at times less than t1"?

Reply:
Yes, it should be `t1 < t2`.